### PR TITLE
[iOS] Custom NSURLSession configuration

### PR DESCRIPTION
**Pull request by hakonk**
## Summary

While it is possible in the React Native implementation for Android to provide a custom configuration for HTTP requests, the iOS implementation does not allow for the same customization. As the NSURLSession used for HTTP requests on iOS is configured internally, one may for instance not supply an ephemeral configuration for HTTP requests. Other concerns related to the given problem have been addressed in the community: https://github.com/react-native-community/discussions-and-proposals/issues/166. I did make a PR with an RFC in the community repo, but after some discussion in the said repo, I figured I might as well make a PR with a suggestion :)

## Changelog

[iOS] [Added] - Allow for configuring the NSURLSessionConfiguration

Implement a C function `RCTSetCustomNSURLSessionConfigurationProvider` which gives the app programmer the ability to provide a block which provides an NSURLSessionConfiguration that will be used for all HTTP requests instead of the default configuration. The provided block will be called when the session configuration is needed.

## Test Plan

Unsure if this can be tested in any other way than uncommenting the example code in `RNTester/RNTester/AppDelegate.mm`.

### CODE REVIEW

**Comment by akshetpandey** [comment]
+1 On this, this will greatly simplify implementations that need to use their own NSURLSessionConfiguration, ex. [RCTCronetHTTPRequestHandler.mm](https://github.com/akshetpandey/react-native-cronet/blob/master/ios/RCTCronetHTTPRequestHandler.mm)

**Comment by zsiegel** [comment]
Any movement on this? I am looking for the ability to customize the iOS networking client like you do on Android.

**Comment by hakonk** [comment]
@zsiegel I haven't heard from the review team. I assume they've had their hands full. Not sure if I should tag someone in the discussion, and in that case, who.

**Comment by safaiyeh** [comment]
Hi @hakonk thanks for the PR, I'll see if I can ping someone for this.

**Comment by hakonk** [comment]
Thanks, @safaiyeh 😊

**Comment by hakonk** [comment]
Hey @safaiyeh, did you have luck finding anyone?

**Comment by safaiyeh** [comment]
Hi @TheSavior, could someone check this out?

**Comment by namchuai** [comment]
Nearly a year already..

**Comment by hakonk** [comment]
Happy new year!

I guess there are more important PRs in the pipe still. Any news @safaiyeh ?

**Comment by pke** [comment]
Interesting that this important (for SSL Pinning that can't use sizzling) PR is still not merged.

File: packages/rn-tester/RNTester/AppDelegate.mm [hunk 1]
```diff
@@ -82,6 +82,12 @@ @implementation AppDelegate
 - (BOOL)application:(__unused UIApplication *)application didFinishLaunchingWithOptions:(NSDictionary *)launchOptions
 {
   RCTEnableTurboModule(YES);
+  // uncomment to configure NSURLSession
```
Comment (TheSavior) [inline packages/rn-tester/RNTester/AppDelegate.mm:85]: We probably don’t need to add commented code here as this is an uncommon use case. Perhaps this explanation of usage should just live in the documentation.

Comment (hakonk) [reply packages/rn-tester/RNTester/AppDelegate.mm:85]: Alright, I can add a section about that. Would it be appropriate to add a section [here](https://reactnative.dev/docs/network) for instance?

Comment (TheSavior) [reply packages/rn-tester/RNTester/AppDelegate.mm:85]: I think so

Comment (hakonk) [reply packages/rn-tester/RNTester/AppDelegate.mm:85]: I have created a PR that contains the needed documentation now: https://github.com/facebook/react-native-website/pull/2629

Comment (hakonk) [reply packages/rn-tester/RNTester/AppDelegate.mm]: Also, the commented code has been removed in: 1634c87

Comment (hakonk) [reply packages/rn-tester/RNTester/AppDelegate.mm]: @TheSavior Please see messages above 🙂

**Comment by taylorkline** [comment]
@pke you mentioned this is useful for SSL pinning.

Can you elaborate on how? From what I can tell, [Apple's docs on performing manual server trust authentication](https://developer.apple.com/documentation/foundation/url_loading_system/handling_an_authentication_challenge/performing_manual_server_trust_authentication) state that we should provide an `NSURLSessionDelegate` that evaluates sever trust within `urlSession(_:didReceive:completionHandler:)`.

This PR does not allow us to set our own `NSURLSessionDelegate`, so how would this be useful in evaluating server trust?

@hakonk definitely open to your input as well. Thank you for this PR.

**Comment by hakonk** [comment]
@taylorkline I believe you are right in that you would have to implement `NSURLSessionDelegate` as a part of implementing certificate or public key pinning, and this PR does not address that particular problem. In `RCTHTTPRequestHandler.mm:99`, you can see that `self` is assigned as the delegate for the `NSURLSession` instance. Thus, there is currently no way of supplying the delegate implementation from outside of React Native.

You can make use of swizzling when you do not have the means to implement `NSURLSessionDelegate` yourself. While I guess this would work most of the time, this mandates potential use of swizzling in other dependencies or your own code. Not really a problem in itself, it just isn't so transparent given that funky stuff happens at runtime. https://github.com/datatheorem/TrustKit does actually allow for implementing public key pinning with swizzling. That could possibly be useful if you want to achieve certificate or public key pinning in a React Native app.

**Comment by taylorkline** [comment]
@hakonk Thank you for confirming and thank you for your time.

**Comment by pke** [comment]
@taylorkline iirc one can only sizzle functions once. So if one react native module sizzles the NSURLSession (for certificate pinning) then no other module could sizzle the session. The ability to provide the pinning configuration for the sessions directly removes the need to use sizzling for it. So thanks for merging this!

**Comment by taylorkline** [comment]
Is this in v0.65.0? The commit does not appear to be tagged with v0.65.0.

**Comment by TheSavior** [comment]
0.65 was cut on April 28th, before this PR was merged. This PR will be included in 0.66

**Comment by hakonk** [comment]
@TheSavior I'm a bit confused. The [changelog](https://github.com/facebook/react-native/pull/27701#issuecomment-900823660) for 0.65.0 lists this PR.

**Comment by TheSavior** [comment]
Ah, I was looking at the tags on the commit as @taylorkline mentioned. It looks like this commit was cherry picked and thus doesn't have the same hash in the release branch. This commit is in 0.65.0 here: https://github.com/facebook/react-native/commit/c299694c87e9912d4c884b5acec04c708ed0f841

Sorry for the confusion! 😅

**Comment by hakonk** [comment]
Alright, thanks for answering so quickly! 😄

**Pull request closed without merging**